server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java

```diff
@@ -250,11 +251,33 @@ public void checkParentLimit(long newBytesReserved, String label) throws Circuit
         long parentLimit = this.parentSettings.getLimit();
         if (totalUsed > parentLimit) {
             this.parentTripCount.incrementAndGet();
-            final String message = "[parent] Data too large, data for [" + label + "]" +
+            final StringBuilder message = new StringBuilder("[parent] Data too large, data for [" + label + "]" +
                     " would be [" + totalUsed + "/" + new ByteSizeValue(totalUsed) + "]" +
                     ", which is larger than the limit of [" +
-                    parentLimit + "/" + new ByteSizeValue(parentLimit) + "]";
-            throw new CircuitBreakingException(message, totalUsed, parentLimit);
+                    parentLimit + "/" + new ByteSizeValue(parentLimit) + "]");
+            if (this.trackRealMemoryUsage) {
+                final long realUsage = currentMemoryUsage();
```

Comment:
Calling `currentMemoryUsage()` again here is tricky because the value is likely different than the one that cause the breaker to trip. To get the correct value, we could calculate it as `final long realUsage = totalUsed - newBytesReserved` which would basically revert the calculation in `#parentUsed(long)` but that also doesn't feel entirely right. Alternatively, we could also return a more structured representation which returns the memory usage as a dedicated property. Wdyt?

Reply:
I thought about this, but I thought that giving the real memory usage at exception construction time was the least-worst option (as part as plain-text is concerned).

For a structured error, I do like the idea of making this exception more structured, it's just a matter of coming up with the structure.

We currently have:

- bytesWanted
- bytesLimit

How about adding

- currentUsage (we could always retrieve it even if "track real memory" is disabled, just for message purposes)
- breakerUsages - a map of the current breakers to their usages

The only downside for this is that we already have logic for serializing the `CircuitBreakingException`, and to make this structured we'd need a new exception (`ParentCircuitBreakingException`) since it doesn't make sense to add this additional structure to every circuit breaker trip (or maybe it does? what do you think?)

Reply:
> For a structured error, I do like the idea of making this exception more structured, it's just a matter of coming up with the structure.

My point was not to change `CircuitBreakingException` but rather to have `#parentUsed(long)` return a more structured representation, something along the lines of:

```java
private static class ParentMemUsage {
    private final long baseUsage;
    private final long reservation;
    private final long totalUsage;
    
    public ParentMemUsage(long baseUsage, long reservation, long totalUsage) {
        // ...
    }
    // ... getters here ...
}

private long parentUsed(long newBytesReserved) {
    if (this.trackRealMemoryUsage) {
        long current = currentMemoryUsage();
        return new ParentMemUsage(current, newBytesReserved,  current + newBytesReserved);
    } else {
        long parentEstimated = 0;
        for (CircuitBreaker breaker : this.breakers.values()) {
            parentEstimated += breaker.getUsed() * breaker.getOverhead();
        }
        return new ParentMemUsage(parentEstimated, newBytesReserved, parentEstimated);
    }
}
```

We could then just call `ParentMemUsage#getBaseUsage()` in order to retrieve the original value again in case an exception occurs. A downside of this is that we allocate an additional object but it might be possible that the compiler is able to optimize it away (by inlining `#parentUsed(long)` and then eliminating the allocation via escape analysis).

Reply:
Okay, I gave this a shot, let me know what you think